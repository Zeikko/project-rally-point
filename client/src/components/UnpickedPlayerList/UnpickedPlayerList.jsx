import PropTypes from 'prop-types'
import React from 'react'
import styled from 'styled-components'
import * as propTypes from '../../constants/prop-types'
import VoteCaptainButton from '../VoteCaptainButton/VoteCaptainButton'
import Player from '../Player/Player'
import PickPlayerButton from '../PickPlayerButton/PickPlayerButton'

function UnpickedPlayersList(props) {
  const {
    players, game, captainVotes, user, dispatch,
  } = props
  if (!players.length) {
    return null
  }
  const unpickedPlayers = players.filter(player => player.team === null)
  return (
    <div>
      <Heading>Players waiting for a game ({unpickedPlayers.length} players)</Heading>
      {unpickedPlayers.map(player => (
        <div key={player.id}>
          <Player player={player} />
          <VoteCaptainButton
            dispatch={dispatch}
            player={player}
            game={game}
            captainVotes={captainVotes}
            user={user}
          />
          <PickPlayerButton
            dispatch={dispatch}
            player={player}
            game={game}
            players={players}
            user={user}
          />
        </div>
      ))}
    </div>
  )
}

UnpickedPlayersList.propTypes = {
  dispatch: PropTypes.func.isRequired,
  game: propTypes.game.isRequired,
  players: PropTypes.arrayOf(propTypes.player).isRequired,
  captainVotes: PropTypes.arrayOf(propTypes.captainVote).isRequired,
  user: propTypes.user.isRequired,
}

export default UnpickedPlayersList

const Heading = styled.div`
  font-size: 20px;
`
