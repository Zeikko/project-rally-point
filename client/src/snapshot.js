import renderer from 'react-test-renderer'
import 'jest-styled-components'

export function matchComponentToSnapshot(component) {
  const json = renderer.create(component).toJSON()
  expect(json).toMatchSnapshot()
}