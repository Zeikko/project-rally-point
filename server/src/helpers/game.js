import _ from 'lodash'
import db from '../db'
import gameStatuses from '../../../common/game-statuses.json'
import playerRoles from '../../../common/player-roles.json'


export async function shouldStartCaptainVote(gameId) {
  const game = await db('game')
    .first('*')
    .where({ id: gameId })
  const numberOfPlayers = await db('player')
    .count('id')
    .where({ gameId })
  return parseInt(numberOfPlayers[0].count, 10) === game.maxPlayers
}

export async function shouldStartSquadLeaderPick(gameId) {
  const game = await db('game')
    .first('*')
    .where({ id: gameId })
  const numberOfCaptainVotes = await db('captainVote')
    .count('id')
    .where({ gameId })
  return parseInt(numberOfCaptainVotes[0].count, 10) === game.maxPlayers
}

export async function shouldStartSquadMemberPick(gameId) {
  const game = await db('game')
    .first('*')
    .where({ id: gameId })
  const numberOfSquadLeaders = await db('player')
    .count('id')
    .where({ gameId, role: playerRoles.SQUAD_LEADER })
  return (parseInt(numberOfSquadLeaders[0].count, 10) + 2) === Math.ceil(game.maxPlayers / 9)
}

export async function startCaptainVote(gameId) {
  const games = await db('game')
    .update({ status: gameStatuses.VOTE_CAPTAINS })
    .where({ id: gameId })
    .returning('*')
  return _.first(games)
}

export async function startSquadLeaderPick(gameId) {
  const games = await db('game')
    .update({
      status: gameStatuses.SQUAD_LEADER_PICK,
      teamWithTurnToPick: 1,
    })
    .where({ id: gameId })
    .returning('*')
  return _.first(games)
}

export async function startSquadMemberPick(gameId) {
  const games = await db('game')
    .update({
      status: gameStatuses.SQUAD_MEMBER_PICK
    })
    .where({ id: gameId })
    .returning('*')
  return _.first(games)
}

export function getGameStatus(gameId) {
  return db('game')
    .first('status')
    .where({ id: gameId })
}

export async function passPlayerPickTurn(id, currentTeam) {
  const games = await db('game')
    .update({
      teamWithTurnToPick: currentTeam === 1 ? 2 : 1
    })
    .where({ id })
    .returning('*')
  return _.first(games)
}
