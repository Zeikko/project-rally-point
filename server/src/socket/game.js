import _ from 'lodash'
import db from '../db'
import actions from '../../../common/actions.json'

export default function handleGame(io, socket, action) {
  switch (action.type) {
    case actions.GET_GAME_REQUEST:
      return getGameRequest(io, socket, action)
    default:
      return null
  }
}

export async function getGameRequest(io, socket) {
  try {
    const game = await db('game')
      .first('*')
      .where({
        status: 'queue',
      })
    socket.emit('action', { type: actions.GET_GAME_SUCCESS, data: game })
  } catch (error) {
    socket.emit('action', { type: actions.GET_GAME_ERROR })
  }
}

export async function shouldStartCaptainVote(gameId) {
  const game = await db('game')
    .first('*')
    .where({ id: gameId })
  const numberOfPlayers = await db('player')
    .count('id')
    .where({ gameId })
  return parseInt(numberOfPlayers[0].count, 10) === game.maxPlayers
}

export async function startCaptainVote(gameId) {
  const games = await db('game')
    .update({ status: 'voting captains' })
    .where({ id: gameId })
    .returning('*')
  return _.first(games)
}

export function createNewGame() {
  return db('game').insert({})
}

export function getGameStatus(gameId) {
  return db('game')
    .first('status')
    .where({ id: gameId })
}

