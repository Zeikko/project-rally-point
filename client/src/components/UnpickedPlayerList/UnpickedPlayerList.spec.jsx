import React from 'react'
import { matchComponentToSnapshot } from '../../test/snapshot'
import UnpickedPlayerList from './UnpickedPlayerList'
import normalUser from '../../../../fixtures/normal-user.json'
import game from '../../../../fixtures/game.json'

describe('UnpickedPlayerList', () => {
  it('renders initial state correctly', () => {
    matchComponentToSnapshot(<UnpickedPlayerList
      dispatch={() => {}}
      gameState={{ game }}
      playersState={{ players: [] }}
      captainVotesState={{ captainVotes: [] }}
      userState={{ isLoading: false, user: null }}
    />)
  })

  it('renders a player correctly', () => {
    matchComponentToSnapshot(<UnpickedPlayerList
      dispatch={() => {}}
      gameState={{ game }}
      playersState={{ players: [normalUser] }}
      captainVotesState={{ captainVotes: [] }}
      userState={{ isLoading: false, user: null }}
    />)
  })
})
