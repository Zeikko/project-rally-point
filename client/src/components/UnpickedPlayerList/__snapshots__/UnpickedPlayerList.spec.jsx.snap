// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnpickedPlayerList renders a player correctly 1`] = `
.c0 {
  font-size: 20px;
}

<div>
  <div
    className="c0"
  >
    Players waiting for a game (
    0
     players)
  </div>
</div>
`;

exports[`UnpickedPlayerList renders initial state correctly 1`] = `null`;
